src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,738 @@
+                {
+                    if (dimension.Key.AnomalyDis.Count > 1)
+                    {
+                        if (best == null || (!Double.IsNaN(valueRatioMap[best]) && (best.AnomalyDis.Count != 1 && (isLeavesLevel ? valueRatioMap[best].CompareTo(dimension.Value) <= 0 : valueRatioMap[best].CompareTo(dimension.Value) >= 0))))
+                        {
+                            best = dimension.Key;
+                        }
+                    }
+                    else
+                    {
+                        if (best == null || best.AnomalyDis.Count > 1)
+                        {
+                            best = dimension.Key;
+                        }
+                        else
+                        {
+                            if (!Double.IsNaN(valueRatioMap[best]) && (isLeavesLevel ? valueRatioMap[best].CompareTo(dimension.Value) <= 0 : valueRatioMap[best].CompareTo(dimension.Value) >= 0))
+                            {
+                                best = dimension.Key;
+                            }
+                        }
+                    }
+                }
+            }
+
+            return best;
+        }
+
+        private double GetSurpriseScore(Point rootCausePoint, Point anomalyPoint)
+        {
+            double p;
+            double q;
+
+            if (anomalyPoint.ExpectedValue == 0)
+            {
+                p = 0;
+            }
+            else
+            {
+                p = rootCausePoint.ExpectedValue / anomalyPoint.ExpectedValue;
+            }
+
+            if (anomalyPoint.Value == 0)
+            {
+                q = 0;
+            }
+            else
+            {
+                q = rootCausePoint.Value / anomalyPoint.Value;
+            }
+
+            double surprise = 0;
+
+            if (p == 0)
+            {
+                surprise = 0.5 * (q * Log2(2 * q / (p + q)));
+            }
+            else if (q == 0)
+            {
+                surprise = 0.5 * (p * Log2(2 * p / (p + q)));
+            }
+            else
+            {
+                surprise = 0.5 * (p * Log2(2 * p / (p + q)) + q * Log2(2 * q / (p + q)));
+            }
+
+            return surprise;
+        }
+
+        private double GetFinalScore(double surprise, double ep, double beta)
+        {
+            double a = 0;
+            double b = 0;
+            if (surprise == 0)
+            {
+                a = 0;
+            }
+            if (ep == 0)
+            {
+                b = 0;
+            }
+            else
+            {
+                a = (1 - Math.Pow(2, -surprise));
+                b = (1 - Math.Pow(2, -ep));
+            }
+
+            return beta * a + (1 - beta) * b;
+        }
+
+        private Tuple<double, double> GetSupriseAndExplainaryScore(Point rootCausePoint, Point anomalyPoint)
```

Comment:
Should the score calculation in some way take AggregateType into account? These make most sense for AggregateType.Sum, right?

Reply:
> Should the score calculation in some way take AggregateType into account? These make most sense for AggregateType.Sum, right?

Yes, has updated